File: togo/internal/services/call_hierarchy.go
package services

// Call hierarchy (incoming / outgoing calls) as pure functions over the program,
// kept separate from the lspserver layer. Port of src/services/callHierarchy.ts.
// Incoming/outgoing re-resolve the method from the item's SelectionRange position
// the client hands back.

import (
	"github.com/nikeee/cappu/internal/compiler"
	"github.com/nikeee/cappu/internal/lsp"
)

const callableFlags = compiler.SymbolFlagsMethod | compiler.SymbolFlagsConstructor

// enclosingCallable returns the method/constructor declaration enclosing a node,
// or nil (a reference in a field initializer / static block has none).
func enclosingCallable(node *compiler.Node) *compiler.Node {
	for n := node.Parent; n != nil; n = n.Parent {
		if n.Kind == compiler.MethodDeclaration || n.Kind == compiler.ConstructorDeclaration {
			return n
		}
	}
	return nil
}

// callHierarchyItemOf builds an item for a method/constructor declaration.
func callHierarchyItemOf(decl *compiler.Node) lsp.CallHierarchyItem {
	file := compiler.GetSourceFileOfNode(decl)
	text := file.AsSourceFile().Text
	lineStarts := compiler.ComputeLineStarts(text)
	name := "<anonymous>"
	kind := lsp.SymbolKindMethod
	if decl.Kind == compiler.ConstructorDeclaration {
		name = "<init>"
		kind = lsp.SymbolKindConstructor
	} else if nm := nodeNameOf(decl); nm != nil && nm.Kind == compiler.Identifier {
		name = nm.AsIdentifier().Text
	}
	selection := decl
	if nm := nodeNameOf(decl); nm != nil {
		selection = nm
	}
	return lsp.CallHierarchyItem{
		Name:           name,
		Kind:           kind,
		URI:            file.AsSourceFile().FileName,
		Range:          dsRange(text, lineStarts, compiler.SkipTrivia(text, decl.Pos), decl.End),
		SelectionRange: dsRange(text, lineStarts, compiler.SkipTrivia(text, selection.Pos), selection.End),
	}
}

func callableSymbolOfItem(program *compiler.Program, checker *compiler.Checker, item lsp.CallHierarchyItem) *compiler.Symbol {
	sourceFile := program.GetSourceFile(compiler.URI(item.URI))
	if sourceFile == nil {
		return nil
	}
	text := sourceFile.AsSourceFile().Text
	lineStarts := compiler.ComputeLineStarts(text)
	offset := compiler.GetPositionOfLineAndCharacter(text, lineStarts, item.SelectionRange.Start.Line, item.SelectionRange.Start.Character)
	id := compiler.GetIdentifierAtPosition(sourceFile, offset)
	if id == nil {
		return nil
	}
	symbol := checker.ResolveName(id)
	if symbol == nil || symbol.Flags&callableFlags == 0 {
		return nil
	}
	return symbol
}

func callableDeclarations(symbol *compiler.Symbol) []*compiler.Node {
	var out []*compiler.Node
	for _, d := range symbol.Declarations {
		if d.Kind == compiler.MethodDeclaration || d.Kind == compiler.ConstructorDeclaration {
			out = append(out, d)
		}
	}
	return out
}

// PrepareCallHierarchy returns the method/constructor at a position.
func PrepareCallHierarchy(checker *compiler.Checker, sourceFile *compiler.Node, offset int) []lsp.CallHierarchyItem {
	id := compiler.GetIdentifierAtPosition(sourceFile, offset)
	if id == nil {
		return nil
	}
	symbol := checker.ResolveName(id)
	if symbol == nil || symbol.Flags&callableFlags == 0 {
		return nil
	}
	var out []lsp.CallHierarchyItem
	for _, d := range callableDeclarations(symbol) {
		out = append(out, callHierarchyItemOf(d))
	}
	return out
}

// CallHierarchyIncoming returns who calls the item's method, grouped by the
// enclosing method/constructor of each call site.
func CallHierarchyIncoming(program *compiler.Program, checker *compiler.Checker, item lsp.CallHierarchyItem) []lsp.CallHierarchyIncomingCall {
	symbol := callableSymbolOfItem(program, checker, item)
	if symbol == nil {
		return nil
	}
	// Preserve first-seen caller order while grouping ranges.
	var order []*compiler.Node
	calls := map[*compiler.Node]*lsp.CallHierarchyIncomingCall{}
	for _, ref := range compiler.FindReferences(symbol, program, checker.ResolveName) {
		if EnclosingCall(ref) == nil {
			continue // only call sites
		}
		caller := enclosingCallable(ref)
		if caller == nil {
			continue
		}
		file := compiler.GetSourceFileOfNode(ref)
		text := file.AsSourceFile().Text
		lineStarts := compiler.ComputeLineStarts(text)
		r := dsRange(text, lineStarts, compiler.SkipTrivia(text, ref.Pos), ref.End)
		if entry, ok := calls[caller]; ok {
			entry.FromRanges = append(entry.FromRanges, r)
		} else {
			calls[caller] = &lsp.CallHierarchyIncomingCall{From: callHierarchyItemOf(caller), FromRanges: []lsp.Range{r}}
			order = append(order, caller)
		}
	}
	var out []lsp.CallHierarchyIncomingCall
	for _, caller := range order {
		out = append(out, *calls[caller])
	}
	return out
}

// CallHierarchyOutgoing returns the callees of the item's method, grouped by
// callee, with the call-site ranges in the method body.
func CallHierarchyOutgoing(program *compiler.Program, checker *compiler.Checker, item lsp.CallHierarchyItem) []lsp.CallHierarchyOutgoingCall {
	symbol := callableSymbolOfItem(program, checker, item)
	if symbol == nil {
		return nil
	}
	var order []*compiler.Node
	calls := map[*compiler.Node]*lsp.CallHierarchyOutgoingCall{}
	for _, decl := range callableDeclarations(symbol) {
		body := declBody(decl)
		if body == nil {
			continue
		}
		file := compiler.GetSourceFileOfNode(decl)
		text := file.AsSourceFile().Text
		lineStarts := compiler.ComputeLineStarts(text)
		var visit func(node *compiler.Node)
		visit = func(node *compiler.Node) {
			if node.Kind == compiler.CallExpression {
				if target := checker.ResolveCall(node); target != nil {
					if nameNode := calleeName(node); nameNode != nil {
						r := dsRange(text, lineStarts, compiler.SkipTrivia(text, nameNode.Pos), nameNode.End)
						if entry, ok := calls[target]; ok {
							entry.FromRanges = append(entry.FromRanges, r)
						} else {
							calls[target] = &lsp.CallHierarchyOutgoingCall{To: callHierarchyItemOf(target), FromRanges: []lsp.Range{r}}
							order = append(order, target)
						}
					}
				}
			}
			node.ForEachChild(func(c *compiler.Node) bool {
				visit(c)
				return false
			})
		}
		visit(body)
	}
	var out []lsp.CallHierarchyOutgoingCall
	for _, callee := range order {
		out = append(out, *calls[callee])
	}
	return out
}

// declBody returns a method/constructor declaration's body, or nil.
func declBody(decl *compiler.Node) *compiler.Node {
	switch decl.Kind {
	case compiler.MethodDeclaration:
		return decl.AsMethodDeclaration().Body
	case compiler.ConstructorDeclaration:
		return decl.AsConstructorDeclaration().Body
	default:
		return nil
	}
}

// calleeName returns the callee name identifier of a call expression, or nil.
func calleeName(call *compiler.Node) *compiler.Node {
	callee := call.AsCallExpression().Expression
	switch callee.Kind {
	case compiler.PropertyAccessExpression:
		return callee.AsPropertyAccessExpression().Name
	case compiler.Identifier:
		return callee
	default:
		return nil
	}
}

File: src/services/callHierarchy.ts
// Call hierarchy (incoming / outgoing calls), as pure functions over the program
// - the transport-free half, mirroring TypeScript's services/callHierarchy and
// cappu's other service modules. The server maps the returned LSP shapes to the
// wire; incoming/outgoing re-resolve the method from the item's selectionRange
// position the client hands back, so nothing opaque has to survive the trip.

import {
  type CallHierarchyIncomingCall,
  type CallHierarchyItem,
  type CallHierarchyOutgoingCall,
  type Range,
  SymbolKind,
} from "vscode-languageserver-types";

import { type Checker } from "../compiler/checker.ts";
import {
  type Character,
  computeLineStarts,
  getLineAndCharacterOfPosition,
  getPositionOfLineAndCharacter,
  type Line,
} from "../compiler/lineMap.ts";
import { forEachChild } from "../compiler/parser.ts";
import type { Program } from "../compiler/program.ts";
import { findReferences, getSourceFileOfNode } from "../compiler/resolver.ts";
import {
  type CallExpression,
  type ConstructorDeclaration,
  type Identifier,
  type MethodDeclaration,
  type Node,
  type PropertyAccessExpression,
  type SourceFile,
  type Symbol,
  SymbolFlags,
  SyntaxKind,
} from "../compiler/types.ts";
import { skipTrivia } from "../compiler/utilities.ts";
import { type Uri } from "../workspace.ts";
import { enclosingCall } from "./hover.ts";
import { getIdentifierAtPosition } from "./nodeAtPosition.ts";

const CALLABLE_FLAGS = SymbolFlags.Method | SymbolFlags.Constructor;

function rangeOf(text: string, lineStarts: readonly number[], pos: number, end: number): Range {
  return {
    start: getLineAndCharacterOfPosition(lineStarts, pos),
    end: getLineAndCharacterOfPosition(lineStarts, end),
  };
}

// The enclosing method or constructor declaration of a node, or undefined (a
// reference in a field initializer / static block has no enclosing callable).
function enclosingCallable(node: Node): MethodDeclaration | ConstructorDeclaration | undefined {
  for (let n: Node | undefined = node.parent; n; n = n.parent) {
    if (n.kind === SyntaxKind.MethodDeclaration) return n as MethodDeclaration;
    if (n.kind === SyntaxKind.ConstructorDeclaration) return n as ConstructorDeclaration;
  }
  return undefined;
}

// A CallHierarchyItem for a method/constructor declaration; selectionRange is the
// name, range spans the declaration.
function itemOfDeclaration(decl: MethodDeclaration | ConstructorDeclaration): CallHierarchyItem {
  const file = getSourceFileOfNode(decl);
  const lineStarts = computeLineStarts(file.text);
  const nameNode = (decl as { name?: { text?: string; pos: number; end: number } }).name;
  const name =
    decl.kind === SyntaxKind.ConstructorDeclaration
      ? "<init>"
      : (nameNode?.text ?? "<anonymous>");
  const selection = nameNode ?? decl;
  return {
    name,
    kind: decl.kind === SyntaxKind.ConstructorDeclaration ? SymbolKind.Constructor : SymbolKind.Method,
    uri: file.fileName,
    range: rangeOf(file.text, lineStarts, skipTrivia(file.text, decl.pos), decl.end),
    selectionRange: rangeOf(file.text, lineStarts, skipTrivia(file.text, selection.pos), selection.end),
  };
}

// The callable symbol an item points at, by re-resolving the identifier at its
// selectionRange start.
function callableSymbolOfItem(program: Program, checker: Checker, item: CallHierarchyItem): Symbol | undefined {
  const sourceFile = program.getSourceFile(item.uri as Uri);
  if (!sourceFile) return undefined;
  const lineStarts = computeLineStarts(sourceFile.text);
  const offset = getPositionOfLineAndCharacter(
    lineStarts,
    item.selectionRange.start.line as Line,
    item.selectionRange.start.character as Character,
  );
  const id = getIdentifierAtPosition(sourceFile, offset) as Identifier | undefined;
  const symbol = id && checker.resolveName(id);
  return symbol && symbol.flags & CALLABLE_FLAGS ? symbol : undefined;
}

function callableDeclarations(symbol: Symbol): (MethodDeclaration | ConstructorDeclaration)[] {
  return (symbol.declarations ?? []).filter(
    d => d.kind === SyntaxKind.MethodDeclaration || d.kind === SyntaxKind.ConstructorDeclaration,
  ) as (MethodDeclaration | ConstructorDeclaration)[];
}

// The method/constructor at a position, as a single-element item list.
export function prepareCallHierarchy(
  checker: Checker,
  sourceFile: SourceFile,
  offset: number,
): CallHierarchyItem[] | null {
  const id = getIdentifierAtPosition(sourceFile, offset) as Identifier | undefined;
  if (!id) return null;
  const symbol = checker.resolveName(id);
  if (!symbol || !(symbol.flags & CALLABLE_FLAGS)) return null;
  const items = callableDeclarations(symbol).map(itemOfDeclaration);
  return items.length > 0 ? items : null;
}

// Who calls the item's method: every call site, grouped by the enclosing
// method/constructor it sits in.
export function callHierarchyIncoming(
  program: Program,
  checker: Checker,
  item: CallHierarchyItem,
): CallHierarchyIncomingCall[] | null {
  const symbol = callableSymbolOfItem(program, checker, item);
  if (!symbol) return null;
  // group caller declaration -> the call-site name ranges within it
  const byCaller = new Map<Node, { from: CallHierarchyItem; ranges: Range[] }>();
  for (const ref of findReferences(symbol, program, checker.resolveName)) {
    if (!enclosingCall(ref as Identifier)) continue; // only call sites, not other uses
    const caller = enclosingCallable(ref);
    if (!caller) continue;
    const file = getSourceFileOfNode(ref);
    const lineStarts = computeLineStarts(file.text);
    const range = rangeOf(file.text, lineStarts, skipTrivia(file.text, ref.pos), ref.end);
    const entry = byCaller.get(caller);
    if (entry) entry.ranges.push(range);
    else byCaller.set(caller, { from: itemOfDeclaration(caller), ranges: [range] });
  }
  const calls = [...byCaller.values()].map(e => ({ from: e.from, fromRanges: e.ranges }));
  return calls.length > 0 ? calls : null;
}

// What the item's method calls: every call in its body, grouped by callee.
export function callHierarchyOutgoing(
  program: Program,
  checker: Checker,
  item: CallHierarchyItem,
): CallHierarchyOutgoingCall[] | null {
  const symbol = callableSymbolOfItem(program, checker, item);
  if (!symbol) return null;
  const byCallee = new Map<Node, { to: CallHierarchyItem; ranges: Range[] }>();
  for (const decl of callableDeclarations(symbol)) {
    const body = (decl as { body?: Node }).body;
    if (!body) continue;
    const file = getSourceFileOfNode(decl);
    const lineStarts = computeLineStarts(file.text);
    const visit = (node: Node): void => {
      if (node.kind === SyntaxKind.CallExpression) {
        const target = checker.resolveCall(node as CallExpression);
        if (target) {
          // the callee name range at this call site
          const callee = (node as CallExpression).expression;
          const nameNode =
            callee.kind === SyntaxKind.PropertyAccessExpression
              ? (callee as PropertyAccessExpression).name
              : callee.kind === SyntaxKind.Identifier
                ? (callee as Identifier)
                : undefined;
          if (nameNode) {
            const range = rangeOf(file.text, lineStarts, skipTrivia(file.text, nameNode.pos), nameNode.end);
            const entry = byCallee.get(target);
            if (entry) entry.ranges.push(range);
            else byCallee.set(target, { to: itemOfDeclaration(target), ranges: [range] });
          }
        }
      }
      forEachChild(node, child => {
        visit(child);
        return undefined;
      });
    };
    visit(body);
  }
  const calls = [...byCallee.values()].map(e => ({ to: e.to, fromRanges: e.ranges }));
  return calls.length > 0 ? calls : null;
}
